pkg/sdn/plugin/dns.go

```diff
@@ -0,0 +1,237 @@
+package plugin
+
+import (
+	"fmt"
+	"net"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	kerrors "k8s.io/kubernetes/pkg/util/errors"
+)
+
+const (
+	DIG = "dig"
+
+	defaultTTL = 30 * time.Minute
+)
+
+type dnsValue struct {
+	// All IPv4 addresses for a given domain name
+	ips []net.IP
+	// Time-to-live value from authoritative name server for the domain
+	ttl time.Duration
+	// Holds (last dns lookup time + ttl), tells when to refresh IPs next time
+	nextQueryTime time.Time
+}
+
+type DNS struct {
+	// Protects dnsMap operations
+	lock sync.Mutex
+	// Holds dns name and its corresponding information
+	dnsMap map[string]dnsValue
+}
+
+func CheckDNSResolver() error {
+	if _, err := exec.LookPath(DIG); err != nil {
+		return fmt.Errorf("%s is not installed", DIG)
+	}
+	return nil
+}
+
+func (d *DNS) Size() int {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	return len(d.dnsMap)
+}
+
+func (d *DNS) Get(dns string) dnsValue {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	data := dnsValue{}
+	if res, ok := d.dnsMap[dns]; ok {
+		data.ips = make([]net.IP, len(res.ips))
+		copy(data.ips, res.ips)
+		data.ttl = res.ttl
+		data.nextQueryTime = res.nextQueryTime
+	}
+	return data
+}
+
+func (d *DNS) Add(dns string) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if len(d.dnsMap) == 0 {
+		d.dnsMap = map[string]dnsValue{}
+	}
+	d.dnsMap[dns] = dnsValue{}
+	d.updateTTL(dns)
+	return d.updateIPs(dns)
+}
+
+func (d *DNS) Update() (bool, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	changed := false
+	errs := []error{}
+	for dns, oldres := range d.dnsMap {
+		if err := d.updateIPs(dns); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if !equal(oldres.ips, d.dnsMap[dns].ips) {
+			changed = true
+		}
+	}
+	return changed, kerrors.NewAggregate(errs)
+}
+
+func (d *DNS) SyncTTL() bool {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	updated := false
+	for dns := range d.dnsMap {
+		if d.updateTTL(dns) {
+			updated = true
+		}
+	}
+	return updated
+}
+
+func (d *DNS) setTTL(dns string, ttl time.Duration) bool {
+	res, ok := d.dnsMap[dns]
+	if !ok {
+		// Should not happen, all operations on dnsMap are synchronized by d.lock
+		glog.Warningf("DNS value not found in dnsMap for domain: %q", dns)
+		return false
+	}
+
+	if res.ttl.Seconds() == ttl.Seconds() {
+		return false
+	}
+	res.ttl = ttl
+	d.dnsMap[dns] = res
+	return true
+}
+
+func (d *DNS) updateTTL(dns string) bool {
+	// Cached/Non-authoritative DNS servers only holds the remaining TTL for it to refresh but not the actual absolute TTL for the domain.
+	// Authoritative DNS servers holds actual TTL for the domain. So we need to:
+	// (a) Get all authorirative name servers for the given domain
+	// (b) Query TTL for the domain against one of the name servers from (a)
```

Comment:
We may not know exact TTL from non-auth server and may query few more times than required with this approach but yes, it definitely reduces the complexity.